arch/src/aarch64/gic.rs

```diff
@@ -54,124 +44,145 @@ pub trait GICDevice: Send + Sync {
-        gic_device: &Box<dyn GICDevice>,
-    ) -> Result<()>
-    where
-        Self: Sized;
-
-    /// Initialize a GIC device
-    fn init_device(vm: &Arc<dyn hypervisor::Vm>) -> Result<DeviceFd>
-    where
-        Self: Sized,
-    {
-        let mut gic_device = kvm_bindings::kvm_create_device {
-            type_: Self::version(),
-            fd: 0,
-            flags: 0,
-        };
-
-        vm.create_device(&mut gic_device).map_err(Error::CreateGIC)
-    }
-
-    /// Set a GIC device attribute
-    fn set_device_attribute(
-        fd: &DeviceFd,
-        group: u32,
-        attr: u64,
-        addr: u64,
-        flags: u32,
-    ) -> Result<()>
-    where
-        Self: Sized,
-    {
-        let attr = kvm_bindings::kvm_device_attr {
-            group: group,
-            attr: attr,
-            addr: addr,
-            flags: flags,
-        };
-        fd.set_device_attr(&attr)
-            .map_err(Error::SetDeviceAttribute)?;
-
-        Ok(())
-    }
-
-    /// Finalize the setup of a GIC device
-    fn finalize_device(gic_device: &Box<dyn GICDevice>) -> Result<()>
-    where
-        Self: Sized,
-    {
-        /* We need to tell the kernel how many irqs to support with this vgic.
-         * See the `layout` module for details.
-         */
-        let nr_irqs: u32 = super::layout::IRQ_MAX - super::layout::IRQ_BASE + 1;
-        let nr_irqs_ptr = &nr_irqs as *const u32;
-        Self::set_device_attribute(
-            gic_device.device_fd(),
-            kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
-            0,
-            nr_irqs_ptr as u64,
-            0,
-        )?;
-
-        /* Finalize the GIC.
-         * See https://code.woboq.org/linux/linux/virt/kvm/arm/vgic/vgic-kvm-device.c.html#211.
-         */
-        Self::set_device_attribute(
-            gic_device.device_fd(),
-            kvm_bindings::KVM_DEV_ARM_VGIC_GRP_CTRL,
-            u64::from(kvm_bindings::KVM_DEV_ARM_VGIC_CTRL_INIT),
-            0,
-            0,
-        )?;
-
-        Ok(())
-    }
-
-    /// Method to initialize the GIC device
-    fn new(vm: &Arc<dyn hypervisor::Vm>, vcpu_count: u64) -> Result<Box<dyn GICDevice>>
-    where
-        Self: Sized,
-    {
-        let vgic_fd = Self::init_device(vm)?;
-
-        let device = Self::create_device(vgic_fd, vcpu_count);
-
-        Self::init_device_attributes(vm, &device)?;
-
-        Self::finalize_device(&device)?;
-
-        Ok(device)
+pub mod kvm {
+    use super::super::gicv2::kvm::KvmGICv2;
+    use super::super::gicv3::kvm::KvmGICv3;
+    use super::super::gicv3_its::kvm::KvmGICv3ITS;
+    use super::super::layout;
+    use super::GICDevice;
+    use super::Result;
+    use kvm_ioctls::DeviceFd;
+    use std::boxed::Box;
+    use std::sync::Arc;
+
+    /// Trait for GIC devices.
+    pub trait KvmGICDevice: Send + Sync + GICDevice {
+        /// Returns the GIC version of the device
+        fn version() -> u32
+        where
+            Self: Sized;
```

Comment:
I know you're only moving code, but is this `Sized` bound really necessary here?

Reply:
Note that you don't need to change this right away in this series. I think it makes more sense to do it separately in a clean up patch or something if it turns out this and other `Sized` bounds are not actually necessary.

Reply:
Indeed, they are not needed. I have removed them.